### ISSUE TITLE
nodesInContext: custom-node handler input grows exponentially with chain length (native OOM) on 0.5x; linear on 1.0.0-beta

### ISSUE THREAD

**Issue by FredrikZeiner** [gorules/zen#512]
## Description

With `overrideConfig({ nodesInContext: true })`, the input payload materialized for a **custom node's** handler grows **exponentially (~2× per upstream passthrough node)**. Behind a realistic chain (we hit this with ~40 nodes), the engine allocates tens of GB of native memory while preparing the custom-node request and the process dies with OOM before the JS handler is ever invoked. Nothing is catchable from JS; in AWS Lambda this kills the invocation.

Pure expression/table graphs are unaffected — the blow-up only happens when the context has to cross the NAPI boundary for a custom node. It looks like each node's stored output embeds its own copy of the `$nodes` context (which embeds the previous node's copy, and so on), and that nested structure is fully materialized for the custom-node request.

## Reproduction (Node.js)

```js
const { ZenEngine, overrideConfig } = require('@gorules/zen-engine');
overrideConfig({ nodesInContext: true });

const N = Number(process.argv[2]);
const nodes = [{ id: 'in', type: 'inputNode', name: 'input', content: { schema: null } }];
const edges = [];
let prev = 'in';
for (let i = 0; i < N; i++) {
  nodes.push({ id: `e${i}`, type: 'expressionNode', name: `expr${i}`, content: { passThrough: true, inputField: null, outputPath: null, executionMode: 'single', expressions: [{ id: `x${i}`, key: `k${i}`, value: '1' }] } });
  edges.push({ id: `ed${i}`, sourceId: prev, targetId: `e${i}`, sourceHandle: null });
  prev = `e${i}`;
}
nodes.push({ id: 'fn', type: 'customNode', name: 'fn', content: { kind: 'functionNode', config: { source: 'x' } } });
nodes.push({ id: 'out', type: 'outputNode', name: 'output', content: { schema: null } });
edges.push({ id: 'ef', sourceId: prev, targetId: 'fn', sourceHandle: null });
edges.push({ id: 'eo', sourceId: 'fn', targetId: 'out', sourceHandle: null });

(async () => {
  const engine = new ZenEngine({ customHandler: async (r) => {
    console.log(`N=${N} custom input bytes=${JSON.stringify(r.input).length}`);
    return { output: { done: true }, traceData: {} };
  } });
  const d = engine.createDecision({ nodes, edges });
  await d.evaluate({ seed: 'ab' }, { trace: false });
  engine.dispose();
})();
```

## Measured custom-handler input size

| chain length N | 0.52.1 / 0.54.0 | 1.0.0-beta.14 |
| --- | --- | --- |
| 3 | 529 B | 175 B |
| 6 | 4 393 B | 367 B |
| 9 | 35 305 B | 622 B |
| 12 | 282 610 B | 947 B |
| 24 | (OOM territory) | 2 975 B |

Stable doubles per node (~2^N); at N≈40 the process allocates multi-GB in seconds and aborts. The 1.0.0 beta line is linear, so this appears already fixed there.

## Ask

Since `latest` (0.54.0) is still exponential and the failure mode is an uncatchable process OOM, could the fix be backported to a 0.54.x patch — or, failing that, could the limitation be documented for `nodesInContext` + custom nodes on the stable line?

## Environment

- `@gorules/zen-engine` 0.52.1 and 0.54.0 (both reproduce), 1.0.0-beta.14 (fixed)
- Node.js 20/22, macOS arm64 and AWS Lambda (arm64, Amazon Linux)
